ghcide/src/Development/IDE/Core/ProgressReporting.hs

```diff
@@ -0,0 +1,229 @@
+    -> IO ProgressReporting
+directProgressReporting sample env style = do
+    st <- newIORef Nothing
+    inProgressVar <- newIORef (HMap.empty @NormalizedFilePath @Int)
+
+    let progressUpdate KickStarted = do
+          u <- newProgressToken
+          writeIORef st (Just u)
+          mRunLspT env $ start u
+        progressUpdate KickCompleted = do
+            mbToken <- atomicSwapIORef st Nothing
+            for_ mbToken $ \u ->
+                mRunLspT env $ stop u
+
+        inProgress file = actionBracket (f file succ) (const $ f file pred) . const
+        -- This function is deliberately eta-expanded to avoid space leaks.
+        -- Do not remove the eta-expansion without profiling a session with at
+        -- least 1000 modifications.
+        f file shift = atomicModifyIORef'_ inProgressVar $
+                HMap.insertWith (\_ x -> shift x) file (shift 0)
+
+        progressLoop :: Double -> LSP.LspM a ()
+        progressLoop prev = do
+            mbToken <- liftIO $ readIORef st
+            case mbToken of
+                Nothing ->
+                    liftIO (sleep sample) >> progressLoop 0
+                Just t -> do
+                    current <- liftIO $ readIORef inProgressVar
+                    prev <- progress style prev current t
+                    liftIO $ sleep sample
+                    progressLoop prev
+
+    progressThread <- async $ mRunLspT env $ progressLoop 0
+    let progressStop = cancel progressThread
+
+    pure ProgressReporting {..}
+
+-- | A 'ProgressReporting' that enqueues Begin and End notifications in a new
+--   thread, with a grace period (nothing will be sent if 'KickCompleted' arrives
+--   before the end of the grace period).
+--   Avoid using in tests where progress notifications are used to assert invariants.
+delayedProgressReporting
+  :: Double  -- ^ sampling rate, also used as grace period before Begin
+  -> Maybe (LSP.LanguageContextEnv c)
+  -> ProgressReportingStyle
+  -> IO ProgressReporting
+delayedProgressReporting sample lspEnv style = do
+    inProgressVar <- newVar (HMap.empty @NormalizedFilePath @Int)
+    mostRecentProgressEvent <- newTVarIO KickCompleted
+    progressAsync <- async $
+        progressThread mostRecentProgressEvent inProgressVar
+    let progressUpdate = atomically . writeTVar mostRecentProgressEvent
+        progressStop   = cancel progressAsync
+        inProgress :: NormalizedFilePath -> Action a -> Action a
+        inProgress = withProgressVar inProgressVar
+    return ProgressReporting{..}
+  where
+    -- The progress thread is a state machine with two states:
+    --   1. Idle
+    --   2. Reporting a kick event
+    -- And two transitions, modelled by 'ProgressEvent':
+    --   1. KickCompleted - transitions from Reporting into Idle
+    --   2. KickStarted - transitions from Idle into Reporting
+    progressThread mostRecentProgressEvent inProgress = progressLoopIdle
+      where
+        progressLoopIdle = do
```

Comment:
I think this does what you are describing, but it's really hard to see with the retries and the running of the reporter async with a kill. Is there a more direct encoding? Like when we get a `progressUpdate Started` we sleep, then spawn the thread, and when we get `Completed` we kill it? I spent a long time trying to find the loop, before realising the loop isn't even in `progressLoop`, it's in `lspShakeProgress`.

Reply:
I'm confused too. There's actually two loops, outer one in the mutually recursive functions forming the state machine, and inner one in `lspShakeProgress`. 

I think the code works fine and I don't see how to make it less confusing, so I've just added more comments.

Reply:
I was expecting the code to read something along the lines of:

```
state <- newVar Nothing
let progressUpdate KickCompleted =
        join $ modifyVar state $ \case
             Nothing -> pure (Nothing, pure ())
             Just thread -> pure (Nothing, cancel thread)
    progressUpdate KickStarted =
        modifyVar_ state $ \case
             Nothing -> Just <$> async (mRunLspT ...)
             Just thread -> pure $ Just thread
```

So you don't need the outer loop. You'd also make progressStop be the same as KickCompleted, simplifying things further.

Reply:
That code is pretty much equivalent to `directProgressReporting` introduced in this PR for tests. And this is a good things, because in hindsight it seems a terrible idea to have one implementation for testing and a different one for production. So the path is clear - just use `directProgressReporting` for both and remove the old version.